Add new Org-level permissions

cneralich opened the pull request.

cneralich committed:
Add new Org-level permissions

cneralich committed:
Add support for selective workspace migration

cneralich committed:
Add workspace_list example file

cneralich committed:
Typo fix

dahlke commented on README.md:
```diff
@@ -71,6 +71,8 @@ By default, the migration tool will load these values from a file named `vcs.jso
 Before initiating the migration process, first determine which command line arguments you wish to pass (if any).  The following arguments are currently supported:
 * `--vcs-file-path`: this flag allows you to pass a custom file path for your TFE_VCS_CONNECTION_MAP JSON file. If not specified, `vcs.json` will be used by default.
 * `--migrate-all-state`: this flag allows you to set the desired behavior for migrating state versions.  If passed, all versions of state will get migrated for all workspaces.  If not specified, only the current version of state for all workspaces will be migrated by default.
+* `--workspace-list-file-path`: this flag allows you to pass a custom file path for your SOURCE_WORKSPACE_LIST TXT file. If not specified, `workspace_list.txt` will be used by default.
```
I would make this a `.json` file. Using `.txt` can be confusing.

dahlke replied:
One thing to note about the issue with the `.txt` if we want to add any additional metadata to this workflow in the future, we'll have to change the format on people at that stage. Might be easier to just start with JSON, but you can also convince me otherwise.

dahlke replied:
One last thing - we may want to call this the `migrate-select-json-path`, since eventually, users will want to use a single input file to select specific resources to migrate, not just workspaces. Might be another reason to use JSON.

cneralich replied:
Good call.  Updated!

dahlke commented on README.md:
```diff
@@ -71,6 +71,8 @@ By default, the migration tool will load these values from a file named `vcs.jso
 * `--vcs-file-path`: this flag allows you to pass a custom file path for your TFE_VCS_CONNECTION_MAP JSON file. If not specified, `vcs.json` will be used by default.
 * `--migrate-all-state`: this flag allows you to set the desired behavior for migrating state versions.  If passed, all versions of state will get migrated for all workspaces.  If not specified, only the current version of state for all workspaces will be migrated by default.
+* `--workspace-list-file-path`: this flag allows you to pass a custom file path for your SOURCE_WORKSPACE_LIST TXT file. If not specified, `workspace_list.txt` will be used by default.
+ * `--migrate-select-workspaces`: this flag allows you to set the desired behavior for migrating workspaces.  If passed, only the workspaces included in your SOURCE_WORKSPACE_LIST TXT file will be migrated.
```
You'll need to update all the all the env vars to have `JSON` instead of `TXT`. Also missing an underscore before `TXT`.

cneralich replied:
Done.

dahlke commented on migration.py:
```diff
@@ -23,21 +24,27 @@
         migrator.migrate_sensitive()
     else:
-        migrator.migrate_all(migrate_all_state, tfe_verify_source)
+        migrator.migrate_all(migrate_all_state, migrate_select_workspaces, tfe_verify_source)
```
I would not overload the `migrate_all` function, and instead, add another function and additional control logic to support this workflow.

dahlke commented on migration.py:
```diff
@@ -23,21 +24,27 @@
+    parser.add_argument('--workspace-list-file-path', dest="workspace_list_file_path", \
+        default=DEFAULT_WORKSPACE_LIST_FILE, \
+            help="Path to the Workspace List file. Defaults to `workspace.txt`.")
+    parser.add_argument('--migrate-select-workspaces', dest="migrate_select_workspaces", action="store_true", \
```
We will likely want to make the `migrate_all` and `migrate_select` workflows be mutually exclusive and provide feedback to the CLI flags so as not to confuse users.

dahlke reviewed (commented):
A couple initial comments on the design for this feature, but overall, looks like it solves the heart of the problem.

cneralich committed:
Genericize optional migration options

cneralich committed:
Update reference

cneralich committed:
Add handling for missing Run Triggers when select items list is provided.

cneralich committed:
Update logging language.

dahlke committed:
- Deduplicating the functions in workspace, handling the select items in a
cleaner way.
- Upgrading to the latest version of terrasnek.
- Adding a TODO file

dahlke committed:
Updating the function name to be `migrate` instead of `migrate_all`
since we are adding the ability to be more selective on what gets
migrated.

dahlke committed:
Fixing some issues with listing resources and deleting resources.

dahlke committed:
Fixing a few other list alls.
